pymc/tests/test_util.py

```diff
@@ -142,3 +143,138 @@ def fn(a=UNSET):
+        (pm.HalfNormal, 0, 1, {"sigma": 1}, {}),
+        (pm.Binomial, 0, 8, {"p": 0.5}, {"n": 10}),
+    ],
+)
+@pytest.mark.parametrize("mass", [0.5, 0.75, 0.95])
+def test_find_optim_prior(distribution, lower, upper, init_guess, fixed_params, mass):
+    with pytest.warns(None) as record:
+        opt_params = pm.find_optim_prior(
+            distribution,
+            lower=lower,
+            upper=upper,
+            mass=mass,
+            init_guess=init_guess,
+            fixed_params=fixed_params,
+        )
+    assert len(record) == 0
+
+    opt_distribution = distribution.dist(**opt_params)
+    mass_in_interval = (
+        pm.math.exp(pm.logcdf(opt_distribution, upper))
+        - pm.math.exp(pm.logcdf(opt_distribution, lower))
+    ).eval()
+    assert np.abs(mass_in_interval - mass) <= 1e-5
+
+
+@pytest.mark.parametrize(
+    "distribution, lower, upper, init_guess, fixed_params",
+    [
+        (pm.Gamma, 0.1, 0.4, {"alpha": 1}, {"beta": 10}),
+        (pm.Exponential, 0.1, 1, {"lam": 1}, {}),
+        (pm.Binomial, 0, 2, {"p": 0.8}, {"n": 10}),
+    ],
+)
+def test_find_optim_prior_error_too_large(distribution, lower, upper, init_guess, fixed_params):
+    with pytest.warns(UserWarning, match="instead of the requested 95%"):
+        pm.find_optim_prior(
+            distribution,
+            lower=lower,
+            upper=upper,
+            mass=0.95,
+            init_guess=init_guess,
+            fixed_params=fixed_params,
+        )
+
+
+def test_find_optim_prior_input_errors():
+    # missing param
+    with pytest.raises(TypeError, match="required positional argument"):
+        pm.find_optim_prior(
+            pm.StudentT,
+            lower=0.1,
+            upper=0.4,
+            mass=0.95,
+            init_guess={"mu": 170, "sigma": 3},
+        )
+
+    # mass too high
+    with pytest.raises(AssertionError, match="has to be between 0.01 and 0.99"):
+        pm.find_optim_prior(
+            pm.StudentT,
+            lower=0.1,
+            upper=0.4,
+            mass=0.995,
+            init_guess={"mu": 170, "sigma": 3},
+            fixed_params={"nu": 7},
+        )
+
+    # mass too low
+    with pytest.raises(AssertionError, match="has to be between 0.01 and 0.99"):
+        pm.find_optim_prior(
+            pm.StudentT,
+            lower=0.1,
+            upper=0.4,
+            mass=0.005,
+            init_guess={"mu": 170, "sigma": 3},
+            fixed_params={"nu": 7},
+        )
+
+    # non-scalar params
+    with pytest.raises(NotImplementedError, match="does not work with non-scalar parameters yet"):
+        pm.find_optim_prior(
+            pm.MvNormal,
+            lower=0,
+            upper=1,
+            mass=0.95,
+            init_guess={"mu": 5, "cov": np.asarray([[1, 0.2], [0.2, 1]])},
+        )
+
+
+distribution = pm.Poisson
+lower = 0
+upper = 10
+if aesara.config.floatX == "float64":
```

Comment:
This is probably due to the discrete nature of the poisson. Can't we pick a range of values where the cdf is more "continuous", like those using high values of mu?

Reply:
Yeah that's exactly why, but I hoped you wouldn't pick up on that 😜 I tried a bunch of values but nothing was really working. I'll give it another try ASAP, because it's not as if today is a special day 🤷‍♂️

Reply:
Haha I'm sorry but you are on the naughty list. Poisson with high mu becomes something more normal-ish, no? I would make the lower bound not be zero anymore.

Reply:
Yeah that's what I'm trying right now. 
Also, this function doesn't work with `pm.Exponential` right now (because the gradient is broken); I think we should use the scipy default when the user provides that distribution. As Harry Potter would say: agreed?

Reply:
Honestly it's a huge pain to find a working test in float32: even `pm.find_optim_prior(pm.Poisson, 93, 107, {"mu": 100}, mass=0.5)` will raise a warning because we get 51% of the mass in there instead of 50%!
Note that this is only a problem for automated tests here, not for a user, who would probably be happy with that result.

I see two options:
- We merge as is
- We introduce a new kwarg that can relax the tolerance of 1% difference between desired and actual mass that we currently have. And we of course relax that assumption for the Poisson tests

Reply:
Yeah lets treat the Poisson separately. Perhaps put it in its own test and test a single mass range that works for both float64 and 32? you already found those right?

Reply:
I think that's what I did in the bit of code we're currently commenting on. Or do you mean something else?

Reply:
I was thinking something that does not need different logic for float64 and float32

Reply:
Ah ok, then I found one combination that works for both floats, but only with mass=0.95. The lower the mass to optimize, the harder

Reply:
Updated